Added ghidra for android reversing

su-vikas opened the pull request:
I have added text for introduction to Ghidra and how to use it for reviewing disassembled code. Some text is same as i wrote for iOS. 

Also fixed few typos in the same document, weren't many to make a new PR, so included in this one. 

This PR covers issue #1366.

cpholguera commented on Document/0x05c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -213,7 +213,7 @@ Most disassemblers can handle any of those architectures. Below, we'll be viewin
 ###### radare2
 
-To open the file in radare2 you only have to run `r2 -A HelloWord-JNI/lib/armeabi-v7a/libnative-lib.so`. The chapter "Android Basic Security Testing" already introduces radare2. Remember that you can use the flag `-A` to run the `aaa` command right after loading the binary in order to _analyze all referenced code_.
+To open the file in radare2 you only have to run `r2 -A HelloWord-JNI/lib/armeabi-v7a/libnative-lib.so`. The chapter "[Android Basic Security Testing](0x05b-Basic-Security_Testing.md/#android-basic-security-testing "Android Basic Security Testing")" already introduced radare2. Remember that you can use the flag `-A` to run the `aaa` command right after loading the binary in order to _analyze all  ...
```
Anchor is not required to refer to the chapter itself :)

```suggestion
To open the file in radare2 you only have to run `r2 -A HelloWord-JNI/lib/armeabi-v7a/libnative-lib.so`. The chapter "[Android Basic Security Testing](0x05b-Basic-Security_Testing.md "Android Basic Security Testing")" already introduced radare2. Remember that you can use the flag `-A` to run the `aaa` command right after loading the binary in order to _analyze all referenced code_.
```

cpholguera commented on Document/0x05c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -394,7 +425,9 @@ $ r2 -qc 'e emu.str=true; s 0x00000e78; af; pdf' HelloWord-JNI/lib/armeabi-v7a/l
-Notice that in this case we're not starting with the `-A` flag not running `aaa`. Instead, we just tell radare2 to analyze that one function by using the _analyze function_ `af` command. This is one fo those cases where we can speed up our workflow because you're focusing on some specific part of an app.
+Notice that in this case we're not starting with the `-A` flag not running `aaa`. Instead, we just tell radare2 to analyze that one function by using the _analyze function_ `af` command. This is one of those cases where we can speed up our workflow because you're focusing on some specific part of an app.
+
+The workflow can further be improved by using [r2ghidra-dec](https://github.com/radareorg/r2ghidra-dec "r2ghidra-dec"), a deep Ghidra decompiler integration for radare2, and generate corresponding decompiled C code. The decompiled C code can aid in quick analysis of the binary.    
```
Very nice remark ;)

```suggestion
The workflow can be further improved by using [r2ghidra-dec](https://github.com/radareorg/r2ghidra-dec "r2ghidra-dec"), a deep integration of Ghidra decompiler for radare2. r2ghidra-dec generates decompiled C code, which can aid in quickly analyzing the binary.    
```

cpholguera commented on Document/0x05c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -444,14 +477,24 @@ BX   R2
+
+###### Ghidra
+
+On opening the library in Ghidra we can see all the function define in the **Symbol Tree** in **Functions** tree. The native library for the current application is relatively very small. There are three user defined functions - `FUN_001004d0`, `FUN_0010051c`, and `Java_sg_vantagepoint_helloworldjni_MainActivity_stringFromJNI`. The other symbols are not user defined and are generated for proper fu ...
```
```suggestion
After opening the library in Ghidra we can see all the functions defined in the **Symbol Tree** panel under **Functions**. The native library for the current application is relatively very small. There are three user defined functions: `FUN_001004d0`, `FUN_0010051c`, and `Java_sg_vantagepoint_helloworldjni_MainActivity_stringFromJNI`. The other symbols are not user defined and are generated for proper functioning of the shared library. The instructions in the function `Java_sg_vantagepoint_helloworldjni_MainActivity_stringFromJNI` are already discussed in detail in previous sections. In this section we can look into the decompilation of the function. 
```

cpholguera commented on Document/0x05c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -444,14 +477,24 @@ BX   R2
+
+On opening the library in Ghidra we can see all the function define in the **Symbol Tree** in **Functions** tree. The native library for the current application is relatively very small. There are three user defined functions - `FUN_001004d0`, `FUN_0010051c`, and `Java_sg_vantagepoint_helloworldjni_MainActivity_stringFromJNI`. The other symbols are not user defined and are generated for proper fu ...
+
+Inside the current function there is a call to another function, whose address is obtained by accessing an offset in JNIEnv pointer. This logic has been diagrammatically demonstrated above as well. Corresponding C code for the disassembled function is shown in the decompilation window. This decompiled C code makes it much easier to understand the function call made. Since this function is small a ...
```
```suggestion
Inside the current function there is a call to another function, whose address is obtained by accessing an offset in the `JNIEnv` pointer (found as `plParm1`). This logic has been diagrammatically demonstrated above as well. The corresponding C code for the disassembled function is shown in the **Decompiler** window. This decompiled C code makes it much easier to understand the function call being made. Since this function is small and extremely simple, the decompilation output is very accurate, this can change drastically when dealing with complex functions.
```

cpholguera replied:
Is the `JNIEnv` pointer explained before? If not could you make the user understand this? This is a very interesting point to highlight when reversing Android native libs. I've just searched the chapter for JNIEnv and it's mentioned but not explained. Maybe you could take the chance to explain it in the first location where it's being referred (around line 180?). That would be very helpful so that the reader can understand what's behind it and why is this offset `0x538` being added to it ;)

su-vikas replied:
Sure, will fix it.

cpholguera commented on Document/0x05c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -510,7 +553,7 @@ In the following section, we'll show how to solve the UnCrackable App for Androi
 ##### Debugging with jdb
 
-The `adb` command line tool was introduced in the "Android Basic Security Testing" chapter. You can use its `adb jdwp` command to list the process ids of all debuggable processes running on the connected device (i.e., processes hosting a JDWP transport). With the `adb forward` command, you can open a listening socket on your host machine and forward this socket's incoming TCP connections to the J ...
+The `adb` command line tool was introduced in the "[Android Basic Security Testing](0x05b-Basic-Security_Testing.md/#android-basic-security-testing "android Basic Security Testing")" chapter. You can use its `adb jdwp` command to list the process ids of all debuggable processes running on the connected device (i.e., processes hosting a JDWP transport). With the `adb forward` command, you can open ...
```
```suggestion
The `adb` command line tool was introduced in the "[Android Basic Security Testing](0x05b-Basic-Security_Testing.md "Android Basic Security Testing")" chapter. You can use its `adb jdwp` command to list the process IDs of all debuggable processes running on the connected device (i.e., processes hosting a JDWP transport). With the `adb forward` command, you can open a listening socket on your host machine and forward this socket's incoming TCP connections to the JDWP transport of a chosen process.
```

su-vikas commented:
Haha. No worries, this issue is a low hanging fruit :)

I second the idea of moving intro to various tools in one place and concentrate on techniques where possible. I opened an issue for this #1501.

cpholguera commented:
Hi, we've agreed on putting a general section in 0x4c.Could you please (as part of this PR) move the Ghidra intro from

- 0x6c : - https://github.com/OWASP/owasp-mstg/blob/master/Document/0x06c-Reverse-Engineering-and-Tampering.md#disassembling-with-ghidra
- and from here (5c)

and make it one subsection within https://github.com/OWASP/owasp-mstg/blob/master/Document/0x04c-Tampering-and-Reverse-Engineering.md#using-disassemblers-and-decompilers ?

Please try to keep it generic in 4c as the goal is to introduce common things in 4c and OS-specific things in 5c and 6c. For instance, what you wrote here fits perfectly the chapter ;) : 
https://github.com/OWASP/owasp-mstg/blob/master/Document/0x06c-Reverse-Engineering-and-Tampering.md#reviewing-disassembled-native-code

Thank you very much! If you have any questions please contact me anytime :)

su-vikas commented:
sure! will update this PR, let me know how it looks then.

su-vikas commented:
@cpholguera Have a look now. Moved the intro to ghidra to 0x04c. Removed the text from other sections. 

 I added some additional info on disassemblers and decompilers in 0x04c  section, feel free to edit/remove it if you find it overkill for an introduction. 

For intro to JNIEnv variable in android, can you open a new ticket against my name. I need to have a look where it will fit in best and already this PR got many things moving around.

And apologies in advance if I am still messing up with the writing guidelines :(.

su-vikas commented:
I have added an intro to JNIEnv. Let me know if it fits in well, it is slightly earlier than what you suggested. 

Also I fixed few sentences, feel free to revert or let me know if they are not apt.

cpholguera commented:
Hi! Thanks for adding it to this PR. It makes it easier for us. I'll give it a review and let you know ;)

Thank you very much!

cpholguera commented on Document/0x04c-Tampering-and-Reverse-Engineering.md:
```diff
@@ -116,9 +116,52 @@ Reverse engineering is the process of reconstructing the semantics of a compiled
+- Position independent code (PIC) sequences, and
+- Hand crafted assembly code.
+
+In similar vain, decompilation is a very complicated process, involving many deterministic and heuristics based approaches. As a consequence, decompilation is not always accurate, but nevertheless very helpful in getting a quick understanding of the function being analyzed. Accuracy of decompilation depends on the amount of information available in the code being decompiled and the sophistication ...
```
```suggestion
On a similar vein, decompilation is a very complicated process, involving many deterministic and heuristic based approaches. As a consequence, decompilation is usually not really accurate, but nevertheless very helpful in getting a quick understanding of the function being analyzed. The accuracy of decompilation depends on the amount of information available in the code being decompiled and the sophistication of the decompiler. In addition, many compilation and post-compilation tools introduce additional complexity to the compiled code in order to increase the difficulty of comprehension and/or even decompilation itself. Such code referred to as _obfuscated code_.
```

cpholguera commented on Document/0x04c-Tampering-and-Reverse-Engineering.md:
```diff
@@ -116,9 +116,52 @@ Reverse engineering is the process of reconstructing the semantics of a compiled
+
+In similar vain, decompilation is a very complicated process, involving many deterministic and heuristics based approaches. As a consequence, decompilation is not always accurate, but nevertheless very helpful in getting a quick understanding of the function being analyzed. Accuracy of decompilation depends on the amount of information available in the code being decompiled and the sophistication ...
+
+Over the past decades many tools have perfected the process of disassembly and decompilation, producing output with high fidelity.  There are a wide range of tools and frameworks available for disassembling and decompiling binaries: commercial tools, open source tools, reverse engineering frameworks, etc. Advanced usage instructions for any of these tools can often easily fill a book of their own ...
```
```suggestion
Over the past decades many tools have perfected the process of disassembly and decompilation, producing output with high fidelity. Advanced usage instructions for any of the available tools can often easily fill a book of their own. The best way to get started is to simply pick up a tool that fits your needs and budget and get a well-reviewed user guide. In this section, we will provide an introduction to some of those tools and in the subsequent "Reverse Engineering and Tampering" Android and iOS chapters we'll focus on the techniques themselves, especially those that are specific to the platform at hand.
```

cpholguera commented on Document/0x04c-Tampering-and-Reverse-Engineering.md:
```diff
@@ -116,9 +116,52 @@ Reverse engineering is the process of reconstructing the semantics of a compiled
+
+##### Ghidra
+
+Ghidra is an open source software reverse engineering (SRE) suite of tools developed by United State of America's National Security Agency's (NSA) Research Directorate. Ghidra is a versatile tool which comprises of a disassembler, decompiler and an inbuilt scripting engine for advance usage. Please refer to the [installation guide](https://ghidra-sre.org/InstallationGuide.html "Ghidra Installatio ...
```
```suggestion
Ghidra is an open source software reverse engineering (SRE) suite of tools developed by the United State of America's National Security Agency's (NSA) Research Directorate. Ghidra is a versatile tool which comprises of a disassembler, decompiler and a built-in scripting engine for advanced usage. Please refer to the [installation guide](https://ghidra-sre.org/InstallationGuide.html "Ghidra Installation Guide") on how to install it and also look at the [cheat sheet](https://ghidra-sre.org/CheatSheet.html "Cheat Sheet") for a first overview of available commands and shortcuts. In this section, we will have walk-through on how to create a project, view disassembly and decompiled code for a binary.
```

cpholguera commented on Document/0x04c-Tampering-and-Reverse-Engineering.md:
```diff
@@ -116,9 +116,52 @@ Reverse engineering is the process of reconstructing the semantics of a compiled
+![Ghidra function graph view](Images/Chapters/0x04c/Ghidra_function_graph.png)
+
+
+There are many other functionalities available in Ghidra and most of them can be explored by opening the **Window** menu. For example, if you want to examine the strings present in the binary, open the **Defined Strings** option. Other advance functionalities we will discuss while analyzing various binaries for Android and iOS platforms in the coming chapters. 
```
```suggestion
There are many other functionalities available in Ghidra and most of them can be explored by opening the **Window** menu. For example, if you want to examine the strings present in the binary, open the **Defined Strings** option. We will discuss other advanced functionalities while analyzing various binaries for Android and iOS platforms in the coming chapters. 
```

cpholguera commented on Document/0x05c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -133,9 +133,11 @@ See the section "[Reviewing Decompiled Java Code](#reviewing-decompiled-java-cod
 Dalvik and ART both support the Java Native Interface (JNI), which defines a way for Java code to interact with native code written in C/C++. As on other Linux-based operating systems, native code is packaged (compiled) into ELF dynamic libraries (\*.so), which the Android app loads at run time via the `System.load` method. However, instead of relying on widely used C libraries (such as glibc), A ...
 
-When reversing Android apps containing native code you'll have to consider this especial layer between Java and native code (JNI). It worths also noticing that when reversing the native code you'll need a disassembler. Once your binary is loaded, you'll be looking at disassembly, which is not _easy_ to look at as Java code.
+For reversing an Android applicaiton containing native code, we need to understand couple of data structures related to JNI bridge between Java and native code. From reversing perspective, we need to be aware of two key data structures, `JavaVM` and `JNIEnv`. Both of these data structures are pointers to pointers to function tables. `JavaVM` provides an interface to invoke functions to create and ...
```
```suggestion
When reversing an Android application containing native code, we need to understand a couple of data structures related to the JNI bridge between Java and native code. From the reversing perspective, we need to be aware of two key data structures: `JavaVM` and `JNIEnv`. Both of them are pointers to pointers to function tables.
- `JavaVM` provides an interface to invoke functions for creating and destroying a JavaVM. Android allows only one `JavaVM` per process and is not really relevant for our reversing purposes.
- `JNIEnv` provides access to most of the JNI functions which are accessible at a fixed offset through the `JNIEnv` pointer. This `JNIEnv` pointer is the first parameter passed to every JNI function. We will discuss this concept again with the help of an example later in this chapter.
```

cpholguera commented on Document/0x05c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -133,9 +133,11 @@ See the section "[Reviewing Decompiled Java Code](#reviewing-decompiled-java-cod
+For reversing an Android applicaiton containing native code, we need to understand couple of data structures related to JNI bridge between Java and native code. From reversing perspective, we need to be aware of two key data structures, `JavaVM` and `JNIEnv`. Both of these data structures are pointers to pointers to function tables. `JavaVM` provides an interface to invoke functions to create and ...
 
-In the next example we'll reverse the HelloWorld-JNI.apk from the OWASP MSTG repository. Installing and running it on your emulator or Android device is optional.
+It is worth highlighting that for reversing the native code in an Android applicaiton we will need a disassembler. And later in the chapter while analyzing the application we will observe that analyzing disassembled native code is more challenging than disassembled Java code.
```
```suggestion
It is worth highlighting that analyzing disassembled native code is much more challenging than disassembled Java code. When reversing the native code in an Android application we will need a disassembler.
```

cpholguera commented on Document/0x05c-Reverse-Engineering-and-Tampering.md:
````diff
@@ -181,7 +183,7 @@ Note the declaration of `public native String stringFromJNI` at the bottom. The
 ```
 
-So where is the native implementation of this function? If you look into the `lib` directory of the APK archive, you'll see eight subdirectories named after different processor architectures. Each of these directories contains a version of the native library `libnative-lib.so` that has been compiled for the processor architecture in question. When `System.loadLibrary` is called, the loader select ...
+So where is the native implementation of this function? If you look into the `lib` directory of the APK archive, you'll see eight subdirectories named after different processor architectures. Each of these directories contains a version of the native library `libnative-lib.so` that has been compiled for the processor architecture in question. When `System.loadLibrary` is called, the loader select ...
````
```suggestion
So where is the native implementation of this function? If you look into the "lib" directory of the unzipped APK archive, you'll see several subdirectories (one per supported processor architecture), each of them containing a version of the native library, in this case `libnative-lib.so`. When `System.loadLibrary` is called, the loader selects the correct version based on the device that the app is running on. Before moving ahead, pay attention to the first parameter passed to the current JNI function. It is the same `JNIEnv` data structure which was discussed earlier in this section.
```

cpholguera commented on Document/0x06c-Reverse-Engineering-and-Tampering.md:
```diff
@@ -56,36 +56,6 @@ If you want to disassemble an application from the App Store, remove the Fairpla
 
 In this section the term "app binary" refers to the Macho-O file in the application bundle which contains the compiled code, and should not be confused with the application bundle - the IPA file. See section "[Exploring the App Package](0x06b-basic-security-testing.md#exploring-the-app-package "Exploring the App Package")" in chapter "Basic iOS Security Testing" for more details on the compositio ...
 
```
General comment: could you please remove all images that are not being used anymore or were relocated, such as: `Document/Images/Chapters/0x05c/Ghidra_import_binary.png`. Thank you!

su-vikas replied:
Fixed.
